README.md
# miner-install

[![Build Status](https://travis-ci.org/postmodern/miner-install.png?branch=master)](https://travis-ci.org/gevans/miner-install)

Forked from [ruby-install](https://github.com/postmodern/ruby-install) and
adapted to work with CGMiner and its myriad of forks.

Currently supports installation of [CGMiner], [BFGMiner], or [SGMiner]

## Features

* Supports installing arbitrary versions.
* Supports installing into `/opt/miners/` for root and `~/.miners/` for users
  by default.
* Supports installing into arbitrary directories.
* Supports downloading from arbitrary URLs.
* Supports downloading from mirrors.
* Supports downloading/applying patches.
* Supports specifying arbitrary `./configure` options.
* Supports downloading archives using `wget` or `curl`.
* Supports verifying downloaded archives using `md5sum`, `md5` or `openssl md5`.
* Supports installing build dependencies via the package manager:
  * [apt]
  * [yum]
  * [pacman]
  * [macports]
  * [brew]
* ~~Has tests.~~ Sorry...

## Anti-Features

* Does not require updating every time a new miner version comes out.
* Does not require recipes for each individual miner version or configuration.
* Does not support installing trunk/HEAD.

## Requirements

* [bash] >= 3.x
* [wget] or [curl]
* `md5sum`, `md5` or `openssl md5`.
* `tar`
* `patch` (if `--patch` is specified)
* [gcc] >= 4.2 or [clang]

## Synopsis

List supported Rubies and their major versions:

    $ miner-install

Install the current stable version of CGMiner:

    $ miner-install cgminer

Install a latest version of SGMiner:

    $ miner-install sgminer 4.1

Install a specific version of BFGMiner:

    $ miner-install bfgminer 3.10.0

Install a miner into a specific directory:

    $ miner-install -i /usr/local/ cgminer 3.7.2

Install a miner with a specific patch:

    $ miner-install -p https://raw.github.com/gist/4136373/falcon-gc.diff cgminer 4.1.0

Install a miner with a specific configuration:

    $ miner-install bfgminer -- --enable-shared --enable-dtrace CFLAGS="-O3"

## System-wide Install

    git clone https://github.com/gevans/miner-install.git
    cd miner-install
    sudo make install

[CGMiner]: https://github.com/ckolivas/cgminer
[SGMiner]: https://github.com/veox/sgminer
[BFGMiner]: https://github.com/luke-jr/bfgminer

[apt]: http://wiki.debian.org/Apt
[yum]: http://yum.baseurl.org/
[pacman]: https://wiki.archlinux.org/index.php/Pacman
[macports]: https://www.macports.org/
[brew]: http://brew.sh

[bash]: http://www.gnu.org/software/bash/
[wget]: http://www.gnu.org/software/wget/
[curl]: http://curl.haxx.se/

[gcc]: http://gcc.gnu.org/
[clang]: http://clang.llvm.org/

[PGP]: http://en.wikipedia.org/wiki/Pretty_Good_Privacy
[1]: http://postmodern.github.com/contact.html#pgp

[homebrew]: http://brew.sh/


CHANGELOG.md
### miner-install 0.1.0

* Forked from [ruby-install](https://github.com/postmodern/ruby-install)

  *Gabe Evans*

* Initial support for CGMiner, SGMiner, and BFGMiner.

  *Gabe Evans*


test/helper.sh
[[ -z "$SHUNIT2"     ]] && SHUNIT2=/usr/share/shunit2/shunit2
[[ -n "$ZSH_VERSION" ]] && setopt shwordsplit

. ./share/miner-install/miner-install.sh
. ./share/miner-install/functions.sh

export PATH="$PWD/bin:$PATH"

function setUp() { return; }
function tearDown() { return; }
function oneTimeTearDown() { return; }


share/miner-install/functions.sh
if (( $UID == 0 )); then
  src_dir="${src_dir:-/usr/local/src}"
  miners_dir="${miners_dir:-/opt/miners}"
else
  src_dir="${src_dir:-$HOME/src}"
  miners_dir="${miners_dir:-$HOME/.miners}"
fi

install_dir="${install_dir:-$miners_dir/$miner-$miner_version}"

#
# Pre-install tasks
#
function pre_install()
{
  mkdir -p "$src_dir" || return $?
  mkdir -p "${install_dir%/*}" || return $?
}

#
# Install miner dependencies
#
function install_deps()
{
  local packages="$(fetch "$miner/dependencies" "$package_manager" || return $?)"

  if [[ -n "$packages" ]]; then
    log "Installing dependencies for $miner $miner_version ..."
    install_packages $packages || return $?
  fi

  install_optional_deps || return $?
}

#
# Install any optional dependencies.
#
function install_optional_deps() { return; }

#
# Download the miner archive
#
function download_miner()
{
  log "Downloading $miner_url into $src_dir ..."
  download "$miner_url" "$src_dir/$miner_archive" || return $?
}

#
# Verifies the miner archive matches a checksum.
#
function verify_miner()
{
  if [[ -n "$miner_md5" ]]; then
    log "Verifying $miner_archive ..."
    verify "$src_dir/$miner_archive" "$miner_md5" || return $?
  else
    warn "No checksum for $miner_archive. Proceeding anyway..."
  fi
}

#
# Extract the miner archive
#
function extract_miner()
{
  log "Extracting $miner_archive ..."
  extract "$src_dir/$miner_archive" "$src_dir" || return $?
}

#
# Initialize ADL SDK variables
#
function init_adl_sdk()
{
  if [ -z "$adl_sdk_version" ]; then
    adl_sdk_version="6.0"
    adl_sdk_md5="66c40ae8b98fd1ad057d2f1a7500d5cd"
  fi

  adl_sdk_archive="${adl_sdk_archive:-adl-sdk-$adl_sdk_version.tar.gz}"
  adl_sdk_src_dir="${adl_sdk_src_dir:-adl-sdk-$adl_sdk_version}"
  adl_sdk_mirror="${adl_sdk_mirror:-http://files.trymining.com/adl-sdk}"
  adl_sdk_url="${adl_sdk_url:-$adl_sdk_mirror/$adl_sdk_archive}"
  miner_adl_sdk_dir="${miner_adl_sdk_dir:-$miner_src_dir/ADL_SDK}"
}

#
# Download the ADL SDK archive
#
function download_adl_sdk()
{
  log "Downloading ADL SDK $adl_sdk_version ..."
  download "$adl_sdk_url" "$src_dir/$adl_sdk_archive" || return $?
}

#
# Verify the ADL SDK archive
#
function verify_adl_sdk()
{
  if [[ -n "$adl_sdk_md5" ]]; then
    log "Verifying ADL SDK $adl_sdk_version ..."
    verify "$src_dir/$adl_sdk_archive" "$adl_sdk_md5" || return $?
  else
    warn "No checksum for $adl_sdk_archive. Proceeding anyway..."
  fi
}

#
# Extract the ADL SDK archive
#
function extract_adl_sdk()
{
  log "Extracting ADL SDK $adl_sdk_version ..."
  extract "$src_dir/$adl_sdk_archive" "$src_dir" || return $?
}

#
# Copy the ADL SDK headers into the miner's source directory
#
function copy_adl_sdk()
{
  log "Copying ADL SDK to $miner_adl_sdk_dir ..."
  mkdir -p "$src_dir/$miner_adl_sdk_dir" || return $?
  shopt -s nullglob
  cp -r "$src_dir/$adl_sdk_src_dir"/* "$src_dir/$miner_adl_sdk_dir/" || return $?
  shopt -u nullglob
}

#
# Install ADL SDK for OpenCL (GPU miners)
#
function install_adl_sdk()
{
  if [[ -z "$miner_install_adl_sdk" ]]; then
    miner_install_adl_sdk=true
  fi

  if [ "$miner_install_adl_sdk" = true ]; then
    init_adl_sdk
    log "Installing ADL SDK $adl_sdk_version ..."
    download_adl_sdk || return $?
    verify_adl_sdk   || return $?
    extract_adl_sdk  || return $?
    copy_adl_sdk     || return $?
  else
    log "Skipping install of ADL SDK."
    return
  fi
}

#
# Download any additional patches
#
function download_patches()
{
  local dest patch

  for (( i=0; i<${#patches[@]}; i++ )) do
    patch="${patches[$i]}"

    if [[ "$patch" == http:\/\/* || "$patch" == https:\/\/* ]]; then
      dest="$src_dir/$miner_src_dir/${patch##*/}"

      log "Downloading patch $patch ..."
      download "$patch" "$dest" || return $?
      patches[$i]="$dest"
    fi
  done
}

#
# Apply any additional patches
#
function apply_patches()
{
  local name

  for patch in "${patches[@]}"; do
    name="${patch##*/}"

    log "Applying patch $name ..."
    patch -p1 -d "$src_dir/$miner_src_dir" < "$patch" || return $?
  done
}

#
# Place holder function for configuring miner.
#
function configure_miner() { return; }

#
# Place holder function for cleaning miner.
#
function clean_miner() { return; }

#
# Place holder function for compiling miner.
#
function compile_miner() { return; }

#
# Place holder function for installing miner.
#
function install_miner() { return; }

#
# Place holder function for post-install tasks.
#
function post_install() { return; }


share/miner-install/miner-install.sh
#!/usr/bin/env bash

shopt -s extglob

miner_install_version="0.1.0"
miner_install_dir="${BASH_SOURCE[0]%/*}"

miners=(cgminer sgminer bfgminer)
patches=()
configure_opts=()
make_opts=()

#
# Auto-detect the package manager.
#
if   command -v apt-get >/dev/null; then package_manager="apt"
elif command -v yum     >/dev/null; then package_manager="yum"
elif command -v port    >/dev/null; then package_manager="port"
elif command -v brew    >/dev/null; then package_manager="brew"
elif command -v pacman  >/dev/null; then package_manager="pacman"
fi

#
# Auto-detect the downloader.
#
if   command -v wget >/dev/null; then downloader="wget"
elif command -v curl >/dev/null; then downloader="curl"
fi

#
# Auto-detect the md5 utility.
#
if   command -v md5sum  >/dev/null; then md5sum="md5sum"
elif command -v md5     >/dev/null; then md5sum="md5"
elif command -v openssl >/dev/null; then md5sum="openssl md5"
fi

#
# Only use sudo if already root.
#
if (( $UID == 0 )); then sudo=""
else                     sudo="sudo"
fi

#
# Prints a log message.
#
function log()
{
  if [[ -t 1 ]]; then
    echo -e "\x1b[1m\x1b[32m>>>\x1b[0m \x1b[1m$1\x1b[0m"
  else
    echo ">>> $1"
  fi
}

#
# Prints a warn message.
#
function warn()
{
  if [[ -t 1 ]]; then
    echo -e "\x1b[1m\x1b[33m***\x1b[0m \x1b[1m$1\x1b[0m" >&2
  else
    echo "*** $1" >&2
  fi
}

#
# Prints an error message.
#
function error()
{
  if [[ -t 1 ]]; then
    echo -e "\x1b[1m\x1b[31m!!!\x1b[0m \x1b[1m$1\x1b[0m" >&2
  else
    echo "!!! $1" >&2
  fi
}

#
# Prints an error message and exists with -1.
#
function fail()
{
  error "$*"
  exit -1
}

#
# Searches a file for a key and echos the value.
# If the key cannot be found, the third argument will be echoed.
#
function fetch()
{
  local file="$miner_install_dir/$1.txt"
  local key="$2"
  local pair="$(grep -E "^$key:" "$file")"

  echo "${pair##$key:*([[:space:]])}"
}

function install_packages()
{
  case "$package_manager" in
    apt)  $sudo apt-get install -y $* || return $? ;;
    yum)  $sudo yum install -y $* || return $?     ;;
    port)   $sudo port install $* || return $?       ;;
    brew)
      local brew_owner="$(/usr/bin/stat -f %Su "$(command -v brew)")"
      sudo -u "$brew_owner" brew install $* ||
      sudo -u "$brew_owner" brew upgrade $* || return $?
      ;;
    pacman)
      local missing_pkgs="$(pacman -T $*)"

      if [[ -n "$missing_pkgs" ]]; then
        $sudo pacman -S $missing_pkgs || return $?
      fi
      ;;
    "")  warn "Could not determine Package Manager. Proceeding anyway." ;;
  esac
}

#
# Downloads a URL.
#
function download()
{
  local url="$1"
  local dest="$2"

  [[ -d "$dest" ]] && dest="$dest/${url##*/}"
  [[ -f "$dest" ]] && return

  case "$downloader" in
    wget) wget -c -O "$dest.part" "$url" || return $?         ;;
    curl) curl -f -L -C - -o "$dest.part" "$url" || return $? ;;
    "")
      error "Could not find wget or curl"
      return 1
      ;;
  esac

  mv "$dest.part" "$dest" || return $?
}

#
# Verifies a file against a md5 checksum.
#
function verify()
{
  local path="$1"
  local md5="$2"

  if [[ -z "$md5sum" ]]; then
    error "Unable to find the md5 checksum utility"
    return 1
  fi

  if [[ -z "$md5" ]]; then
    error "No md5 checksum given"
    return 1
  fi

  if [[ "$($md5sum "$path")" != *$md5* ]]; then
    error "$path is invalid!"
    return 1
  fi
}

#
# Extracts an archive.
#
function extract()
{
  local archive="$1"
  local dest="${2:-${archive%/*}}"

  case "$archive" in
    *.tgz|*.tar.gz) tar -xzf "$archive" -C "$dest" || return $? ;;
    *.tbz|*.tbz2|*.tar.bz2)  tar -xjf "$archive" -C "$dest" || return $? ;;
    *.zip) unzip "$archive" -d "$dest" || return $? ;;
    *)
      error "Unknown archive format: $archive"
      return 1
      ;;
  esac
}

#
# Loads function.sh for the given miner.
#
function load_miner()
{
  miner_dir="$miner_install_dir/$miner"

  if [[ ! -d "$miner_dir" ]]; then
    echo "miner-install: unsupported miner: $miner" >&2
    return 1
  fi

  local expanded_version="$(fetch "$miner/versions" "$miner_version")"
  miner_version="${expanded_version:-$miner_version}"

  source "$miner_install_dir/functions.sh" || return $?
  source "$miner_dir/functions.sh" || return $?

  miner_md5="${miner_md5:-$(fetch "$miner/md5" "$miner_archive")}"
}

#
# Prints miners supported by miner-install.
#
function known_miners()
{
  echo "Known miner versions:"

  for miner in ${miners[@]}; do
    echo "  $miner:"
    cat "$miner_install_dir/$miner/versions.txt" | sed -e 's/^/    /' || return $?
  done
}

#
# Prints usage information for miner-install.
#
function usage()
{
  cat <<USAGE
usage: miner-install [OPTIONS] [MINER [VERSION] [-- CONFIGURE_OPTS ...]]

Options:

  -M, --mirror URL   Alternate mirror to download the miner archive from
  -i, --install-dir  DIR  Directory to install miner into
  -j, --jobs JOBS    Number of jobs to run in parallel when compiling
  -m, --md5 MD5      MD5 checksum of the miner archive
  -p, --patch FILE   Patch to apply to the miner source code
  -r, --miners-dir   DIR  Directory that contains other installed miners
  -s, --src-dir DIR  Directory to download source-code into
  -u, --url URL      Alternate URL to download the miner archive from
  --no-download      Use the previously downloaded miner archive
  --no-install-deps  Do not install build dependencies before installing miner
  --reinstall        Allow installation if another miner is detected in same location
  --no-verify        Do not verify the downloaded miner archive
  -V, --version      Prints the version
  -h, --help         Prints this message

Examples:

  $ miner-install sgminer
  $ miner-install sgminer 4.1
  $ miner-install sgminer 4.1.0
  $ miner-install sgminer -- --with-openssl-dir=...
  $ miner-install -M https://mirror.example.com/pub/sgminer sgminer
  $ miner-install -p https://raw.github.com/gist/4136373/falcon-gc.diff cgminer 3.7.2

USAGE
}

#
# Parses command-line options for miner-install.
#
function parse_options()
{
  local argv=()

  while [[ $# -gt 0 ]]; do
    case $1 in
      -r|--miners-dir)
        miners_dir="$2"
        shift 2
        ;;
      -i|--install-dir)
        install_dir="$2"
        shift 2
        ;;
      -s|--src-dir)
        src_dir="$2"
        shift 2
        ;;
      -j|--jobs|-j+([0-9])|--jobs=+([0-9]))
        make_opts+=("$1")
        shift
        ;;
      -p|--patch)
        patches+=("$2")
        shift 2
        ;;
      -M|--mirror)
        miner_mirror="$2"
        shift 2
        ;;
      -u|--url)
        miner_url="$2"
        shift 2
        ;;
      -m|--md5)
        miner_md5="$2"
        shift 2
        ;;
      --no-download)
        no_download=1
        shift
        ;;
      --no-verify)
        no_verify=1
        shift
        ;;
      --no-install-deps)
        no_install_deps=1
        shift
        ;;
      --reinstall)
        reinstall=1
        shift
        ;;
      -V|--version)
        echo "miner-install: $miner_install_version"
        exit
        ;;
      -h|--help)
        usage
        exit
        ;;
      --)
        shift
        configure_opts=("$@")
        break
        ;;
      -*)
        echo "miner-install: unrecognized option $1" >&2
        return 1
        ;;
      *)
        argv+=($1)
        shift
        ;;
    esac
  done

  case ${#argv[*]} in
    2)
      miner="${argv[0]}"
      miner_version="${argv[1]}"
      ;;
    1)
      miner="${argv[0]}"
      miner_version="stable"

      local miner_dir="$miner_install_dir/$miner"
      if [[ ! -d "$miner_dir" ]]; then
        miner="${argv[0]%%-*}"
        miner_version="${argv[0]#*-}"
      fi
      ;;
    0)
      echo "miner-install: too few arguments" >&2
      usage 1>&2
      return 1
      ;;
    *)
      echo "miner-install: too many arguments: ${argv[*]}" >&2
      usage 1>&2
      return 1
      ;;
  esac
}
